### PR TITLE
Granite speech speedups

### PR DESCRIPTION
**Pull request by avihu111**
# What does this PR do?
This PR speeds up Granite speech by the following changes:
1. Register attention_dist to avoid cpu-to-gpu transfer every layer.
2. `pad_sequence` is much faster than per-sample-padding + concat.
3. Avoid returning audio to cpu when using a compute device.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@ArthurZucker @eustlb can you give that a look 🙏 
CC: @gsaon @alex-jw-brooks @avishaiElmakies

### CODE REVIEW

File: src/transformers/models/granite_speech/feature_extraction_granite_speech.py [hunk 1]
```diff
@@ -178,11 +176,8 @@ def _get_audios_and_audio_lengths(self, audios: AudioInput) -> Sequence["torch.T
             if not torch.is_floating_point(audios[0]):
                 raise ValueError("Invalid audio provided. Audio should be a floating point between 0 and 1")
             lengths = [audio.shape[-1] for audio in audios]
-            padding = [max(lengths) - length for length in lengths]
-            # ensure all audios have a batch dimension:
-            audios = [audio.view(1, -1) for audio in audios]
-            padded = [torch.nn.functional.pad(audio, (0, pad)) for audio, pad in zip(audios, padding)]
-            audios = torch.cat(padded, dim=0)
+            audios = [audio.squeeze(0) for audio in audios]
```
Comment (avihu111) [inline src/transformers/models/granite_speech/feature_extraction_granite_speech.py:179]: `pad_sequence` is much faster

File: src/transformers/models/granite_speech/modeling_granite_speech.py [hunk 2]
```diff
@@ -258,8 +257,8 @@ def __init__(self, config: GraniteSpeechEncoderConfig):
         # Precompute clamped relative positional encoding distances
         seq = torch.arange(config.context_size)
         relpos_dist = seq.view(-1, 1) - seq.view(1, -1)
-        self.attention_dists = torch.clamp(relpos_dist, -config.context_size, config.context_size) + config.max_pos_emb
-
+        attention_dists = torch.clamp(relpos_dist, -config.context_size, config.context_size) + config.max_pos_emb
+        self.register_buffer("attention_dists", attention_dists, persistent=False)
```
Comment (avihu111) [inline src/transformers/models/granite_speech/modeling_granite_speech.py:261]: registering the buffer automatically passes this to cuda, no need to re-move from cpu to gpu every forward pass

File: src/transformers/models/granite_speech/feature_extraction_granite_speech.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/granite_speech/feature_extraction_granite_speech.py:179]: wow that's good to know! How much faster is it exactly?

File: src/transformers/models/granite_speech/modeling_granite_speech.py (hunk 2 above)
Comment (ArthurZucker) [reply src/transformers/models/granite_speech/modeling_granite_speech.py:261]: that is a good find!

**Review by ArthurZucker — APPROVED** [review]
Thanks a lot for continuing to improve the speed of the model!

**Comment by ArthurZucker** [comment]
Could you give some numbers about the speedup?

File: src/transformers/models/granite_speech/feature_extraction_granite_speech.py (hunk 1 above)
Comment (avihu111) [reply src/transformers/models/granite_speech/feature_extraction_granite_speech.py:179]: @ArthurZucker I was surprised too. 

I think there are two reasons for that:
- The previous code required multiple memory allocations. (Each padding requires a new memory allocation, as well as the concatenation.)
- `pad_sequence` is compiled. 

I ran it for 500 repeats:
```
old took 2.545 seconds
new took 1.096 seconds
```

It gave large utilization improvements when using the model in `trl`.
Also, the code looks cleaner. :)

File: src/transformers/models/granite_speech/modeling_granite_speech.py (hunk 2 above)
Comment (avihu111) [reply src/transformers/models/granite_speech/modeling_granite_speech.py:261]: @ArthurZucker sharing the timing of old vs new code (just those lines, 500 repeats) 
```
old took 0.029 seconds
new took 0.005 seconds
```

**Comment by avihu111** [comment]
Sure! Thanks for looking into this! Commented on each part separately 👍

**Comment by ArthurZucker** [comment]
Thanks a lot, that's a great contribution for audio! 🤗
